[Core] Add Reboot keycode to core

drashna opened the pull request:
Adds a reboot function and keycode to core.  This allows the board to be reset without unplugging it.

Have tested on non-VUSB AVR, ATSAM, and on STM32 DFU. 

## Additional Behaviors 
* `reset_keyboard` jumps to bootloader, behavior not changed.
* everything but bootloader jump removed from `reset_keyboard()` function and moved to a separate function that is called by `reset_keyboard()`
  * This allows for the shutdown behavior to be called from both the bootloader jump code and the mcu reset code. 
* `soft_reset_keyboard` re-inits board by calling mcu_reset
* EEPROM reset calls `soft_reset_keyboard()` to properly re-init all settings
  * This (already) includes a hard 250ms wait after `eeprom_init()`, which should ensure eeprom is properly flushed
* Adds `QK_REBOOT` keycode that calls the soft reset.

## Caveats
* gd32v dfu - no way to jump to bootloader, bootloader jump and mcu reset perform same function

These bootloader may need user cleanup after failed reboots.

## Types of Changes

- [x] Core
- [x] New feature
- [x] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

tzarc commented:
Shall we rename it to `mcu_reset()` instead of `keyboard_jump()`? Jumping to keyboard kinda makes me think it should go back to "normal keyboard duties" which seems a bit weird.

sigprof commented on platforms/arm_atsam/bootloaders/md_boot.c:
```diff
@@ -38,17 +38,21 @@ void bootloader_jump(void) {
     if (!*ver_check) {                   // If check version pointer is NULL, all characters have matched
         *MAGIC_ADDR = BOOTLOADER_MAGIC;  // Set magic number into RAM
-        NVIC_SystemReset();              // Perform system reset
+        keyboard_jump();              // Perform system reset
```
Looks like most (but not all) implementations of `bootloader_jump()` now use `keyboard_jump()` instead of performing system reset directly; however, `keyboard_jump()` is weak and therefore can be replaced — but with this implementation there is no way to override just the “restart keyboard” function without also affecting the “reboot to bootloader” function. Is this actually intended? Maybe it would be better to add a separate `mcu_reset()` function and use it in both `bootloader_jump()` and `keyboard_jump()` (although there is probably no point doing that for just `NVIC_SystemReset()`, some MCUs need much more elaborate code there).

Also it might make sense to write something invalid into the bootloader magic location for the “restart keyboard” case, so that the restart happens even if something else happened to be using that RAM location. This would also require having a separate “restart keyboard” function.

sigprof commented on platforms/arm_atsam/bootloaders/md_boot.c:
```diff
@@ -69,4 +73,11 @@ void bootloader_jump(void) {
         ;  // Wait on timeout
 }
+
+__attribute__((weak)) void keyboard_jump(void) {
+    WDT->CONFIG.bit.WINDOW   = 0;
+    WDT->CONFIG.bit.PER      = 0;
+    WDT->EWCTRL.bit.EWOFFSET = 0;
+    WDT->CTRLA.bit.ENABLE    = 1;
+}
```
Looks like this implementation is incomplete — it starts the watchdog timer, but does not enter the infinite loop, so it might return and do something undesired while the watchdog timer is counting down.

drashna replied:
Honestly, I wasn't entirely sure about this, and I don't have hardware to test on.

sigprof commented on platforms/chibios/bootloaders/none.c:
```diff
@@ -19,5 +19,6 @@
 #pragma message "Unknown bootloader set, you may not be able to enter bootloader using software reset"
 
 __attribute__((weak)) void bootloader_jump(void) {}
+__attribute__((weak)) void keyboard_jump(void) { NVIC_SystemReset(); }
```
This might have to remain empty, or use `#ifdef NVIC_SystemReset`, if we want to keep this file buildable for any weird MCU (or even not so weird — GD32V does not have that function, although normally it should use its own file).

sigprof commented on platforms/chibios/bootloaders/stm32duino.c:
```diff
@@ -18,4 +18,5 @@
 #include <ch.h>
 
-__attribute__((weak)) void bootloader_jump(void) { NVIC_SystemReset(); }
+__attribute__((weak)) void bootloader_jump(void) { keyboard_jump(); }
+__attribute__((weak)) void keyboard_jump(void) { NVIC_SystemReset(); }
```
For stm32duino the bootloader handling in QMK is somewhat weird — the bootloader magic word is in the RTC backup RAM, and it gets set to the “enter the bootloader” value in the QMK startup code (so that pressing Reset just once after the QMK firmware had started will result in entering the bootloader for the unlimited time). Here `keyboard_jump()` may need to clear that magic word, so that the keyboard restart will actually work without getting stuck in the bootloader (especially if the “fastboot” version of the bootloader is used).

drashna commented:
For now, I've renamed `keyboard_jump` to `mcu_reset`.  Additionally, I've reverted changes that call mcu_reset from bootloader_jump, for now. 

I've also reverted the massdrop and stm32duino code, since it's unclear, ATM.


I prefer having a reset function that is called by both options based on what is needed, especially as a lot of this is repeated code (watchdog timers, or nvic_systemreset).

KarlK90 commented on platforms/chibios/bootloaders/gd32v_dfu.c:
```diff
@@ -36,5 +36,9 @@ __attribute__((weak)) void bootloader_jump(void) {
     *DBGMCU_CMD = DBGMCU_CMD_RESET;
 }
 
+__attribute__((weak)) void mcu_reset(void) {
```
You have to unlock the dbg registers first, and after that issue the command. basically what jump to bootloader does.

drashna replied:
Is there a way to just reset the firmware? 

And I'm guessing that `NVIC_SystemReset()` won't work for this, correct?

drashna replied:
Wait, does that mean that the reset keycode/etc will just reboot the MCU, but not jump to the bootloader? 

or I am just misunderstanding? 

Eg, the end goal is to have a way to jump to bootloader, and a way to reset the mcu, to re-init everything.

sigprof replied:
Does resetting to the bootloader work on GD32V if you implement a reset circuit like the one which is needed for dual-bank STM32 chips (a GPIO is used to charge a capacitor on BOOT0)?

KarlK90 replied:
> Wait, does that mean that the reset keycode/etc will just reboot the MCU, but not jump to the bootloader?

This is correct https://github.com/Nuclei-Software/nuclei-sdk/issues/29 has all the gory details and failed attempts on how to enter the bootloader. 

This is also the reason why I ported tinyuf2 to the gd32vf103 series, but haven't had the opportunity to finish that (it works but needs to be merged into tinyuf2 and qmk).

KarlK90 replied:
> Does resetting to the bootloader work on GD32V if you implement a reset circuit like the one which is needed for dual-bank STM32 chips (a GPIO is used to charge a capacitor on BOOT0)?

I tried that both with the reset circuit from the djinn and another one previously posted by ebaster in the qmk discord. (But not the acheron project ones). Both did not work at all, I wouldn't rule out that some other modification could make it work. At this point tinyuf2 is the better choice, or simple dedicated boot0 and reset buttons if one can live with that.

drashna replied:
SO, for the mcu_reset, what would the best behavior be? 
Just leaving it blank/empty with no behavior? 

Or would the `NVIC_SystemReset()` function still work here to reset the firmware?

KarlK90 replied:
mcu_reset should have the same code as jump to bootloader. It just issues a reset nothing more.

zvecr commented on quantum/quantum.c:
```diff
@@ -330,6 +339,10 @@ bool process_record_quantum(keyrecord_t *record) {
                 eeconfig_init();
+                soft_reset_keyboard();
+                return false;
+            case QK_REBOOT:
```
This block should probably be moved under `QK_BOOTLOADER` within the `#ifndef NO_RESET`

zvecr commented on platforms/arm_atsam/bootloaders/md_boot.c:
```diff
@@ -63,3 +63,7 @@ void bootloader_jump(void) {
         ; // Wait on timeout
 }
+
+__attribute__((weak)) void mcu_reset(void) {
+    // not certain how to setup wdt here
+}
```
Kinda minor, but you cant use watchdog, as its use will trigger the device to stay in the bootloader
```suggestion
__attribute__((weak)) void mcu_reset(void) {}
```

tzarc replied:
Will see if `NVIC_SystemReset()` works.

tzarc replied:
Works fine.
```suggestion
__attribute__((weak)) void mcu_reset(void) {
    NVIC_SystemReset();
}
```
